Migration Guide - Create React App (#3788)
withastro/docs: src/content/docs/en/guides/migrate-to-astro/from-create-react-app.mdx

You can pass a `--template` argument to the `create astro` command to start a new Astro project with one of our official starters (e.g. `docs`, `blog`, `portfolio`). Or, you can [start a new project from any existing Astro repository on GitHub](/en/install/auto/#starter-templates).

 <PackageManagerTabs>
    <Fragment slot="npm">
    ```shell
    # launch the Astro CLI Wizard 
    npm create astro@latest

    # create a new project with an official example
    npm create astro@latest -- --template <example-name>
    ```
    </Fragment>
    <Fragment slot="pnpm">
    ```shell
    # launch the Astro CLI Wizard 
    pnpm create astro@latest

    # create a new project with an official example
    pnpm create astro@latest --template <example-name>
    ```
    </Fragment>
    <Fragment slot="yarn">
    ```shell
    # launch the Astro CLI Wizard 
    yarn create astro@latest

    # create a new project with an official example
    yarn create astro@latest --template <example-name>
    ```
    </Fragment>
  </PackageManagerTabs>

> Then, copy your existing Creat React App project files over to your new Astro project in a separate folder outside of `src`.

Reviewer: typo
```suggestion
Then, copy your existing Create React App project files over to your new Astro project in a separate folder outside of `src`.
```

:::tip
Visit https://astro.new for the full list of official starter templates, and links for opening a new project in StackBlitz, CodeSandbox, or Gitpod.
:::

### Install integrations (optional)

You may find it useful to install some of [Astro's optional integrations](/en/guides/integrations-guide/) to use while converting your Create React App project to Astro. For example, **@astrojs/react** lets you reuse some existing React UI components in your new Astro site, or keep writing with React components.

### Put your source code in `src`

Following [Astro's project structure](/en/core-concepts/project-structure/):

1. **Keep** Create React App's `public/` folder untouched. 
   
    Astro uses the `public/` directory for static assets, just like Create React App. There is no change needed to this folder, nor its contents.

2. **Copy or Move** Create React App's other files and folders (e.g. `components`, `hooks`, `styles`, etc.) into Astro's `src/` folder as you rebuild your site, following [Astro's project structure](/en/core-concepts/project-structure/).

>     Astro's `src/pages/` folder is a special folder used for file-based routing. All other folders are optional. The contents of your `src/` folder can be organized any way. You can also include common folders in Astro projects include `src/layouts/`, `src/components`, `src/styles`, `src/scripts`.

Reviewer: `any way` -> `anyway`
```suggestion
    Astro's `src/pages/` folder is a special folder used for file-based routing. All other folders are optional. The contents of your `src/` folder can be organized anyway. You can also include common folders in Astro projects include `src/layouts/`, `src/components`, `src/styles`, `src/scripts`.
```

Note that `.astro` and several other file types must be imported with their full file extension.

```astro title="src/pages/authors/Fred.astro"
---
import Card from `../../components/Card.astro`
---
<Card />
```

### Create React App Children Props to Astro

Convert any instances of `{children}` to an Astro `<slot />`. Astro does not need to receive `{children}` as a function prop and will automatically render child content in a `<slot />`.

```astro title="src/components/MyComponent" del={3-9} ins={11-13}
---
---
export default function MyComponent(props) { 
    return (
      <div>
        {props.children}
      </div>
    );  
}

<div>
  <slot />
</div>
```

React components that pass multiple sets of children can be migrated to an Astro component using [named slots](/en/core-concepts/astro-components/#named-slots).

See more about [specific `<slot />` usage in Astro](/en/core-concepts/astro-components/#slots).

### Create React App Data Fetching to Astro

Fetching data in a Create React App component is similar to Astro, with some slight differences.

> You will need to remove any instances of a side effect hook (`useEffect`) for either `Astro.glob()` or `getCollection()`/`getEntryBySlug()` to access data other files in your project source. To [fetch remote data](/en/guides/data-fetching/), use `fetch()`.

Reviewer: Missing `from`
```suggestion
You will need to remove any instances of a side effect hook (`useEffect`) for either `Astro.glob()` or `getCollection()`/`getEntryBySlug()` to access data from other files in your project source. To [fetch remote data](/en/guides/data-fetching/), use `fetch()`.
```

---
title: Migrating from Create React App
description: Tips for migrating an existing Create React App project to Astro
type: migration
stub: true
framework: Create React App
i18nReady: true
---
import AstroJSXTabs from '~/components/tabs/AstroJSXTabs.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

> 

Reviewer 2: ```suggestion
import FileTree from `~components/FileTree.astro'
```

---
title: Migrating from Create React App (CRA)
description: Tips for migrating an existing Create React App project to Astro
type: migration
stub: true
framework: Create React App
i18nReady: true
---
import AstroJSXTabs from '~/components/tabs/AstroJSXTabs.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import FileTree from '~/components/FileTree.astro'

Astro's [React integration](/en/guides/integrations-guide/react/) provides support for [using React components inside Astro components](/en/core-concepts/framework-components/), including entire React apps like Create React App (CRA)!

```astro title="src/pages/index.astro"
---
// Import your root App component

> import App from '../cra-project/App.jsx'

Reviewer 2: ```suggestion
import App from '../cra-project/App'
```

---
<!-- Use a client directive to load your app -->

> <App client:load/>

Reviewer 2: ```suggestion
<App client:load />
```

```

> Many apps will "just work" as full React apps when you add them directly to your Astro project with the React integration installed. This is a great way to get your project up and running immediately, and keep your app functional while you migrate to Astro. 

Reviewer: ```suggestion
Many apps will "just work" as full React apps when you add them directly to your Astro project with the React integration installed. This is a great way to get your project up and running immediately and keep your app functional while you migrate to Astro.
```

Here are some key concepts and migration strategies to help you get started. Use the rest of our docs and our [Discord community](https://astro.build/chat) to keep going!

## Key Similarities between Create React App and Astro

- The [syntax of `.astro` files is similar to JSX](/en/core-concepts/astro-syntax/#differences-between-astro-and-jsx). Writing Astro should feel familiar.

- Astro uses file-based routing, and [allows specially named pages to create dynamic routes](/en/core-concepts/routing/#dynamic-routes).

- Astro is [component-based](/en/core-concepts/astro-components/), and your markup structure will be similar before and after your migration.

- Astro has [official integrations for React, Preact, and Solid](/en/guides/integrations-guide/react/) so you can use your existing JSX components. Note that in Astro, these files **must** have a `.jsx` or `.tsx` extension.

- Astro has support for [installing NPM packages](/en/guides/imports/#npm-packages), including React libraries. Many of your existing dependencies will work in Astro.

## Key Differences between Create React App and Astro

> When you rebuild your Create React App site in Astro, you will notice some important differences:

Author: Could we possibly introduce the CRA acronym here or is it more appropriate later on? 

I was also curious of what your thoughts would be having `(CRA)` in title? That way you save the writing in parts such as `Key Difference between CRA and Astro`, as an example.

Reviewer 2: Yes, those would make for nicer side bar headings. I could see putting it in the title, and maybe even also in the opening line? Let's try both places and then use the acronym anywhere the long name absolutely feels too long (esp. headings, for sidebar real estate.)

## Add your Create React App to Astro

Your existing app can be rendered directly inside a new Astro project, often with no changes to your app's code.

### Create a new Astro project

Use the `create astro` command for your package manager to launch Astro's CLI wizard and select a new "empty" Astro project.

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm create astro@latest
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm create astro@latest
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn create astro@latest
  ```
  </Fragment>
</PackageManagerTabs>

### Add integrations and dependencies
Add the React integration using the `astro add` command for your package manager. If your app uses Tailwind or MDX, you can add multiple Astro integrations using the same command:

```sh
# Using NPM
npx astro add react
npx astro add react tailwind mdx
# Using Yarn
yarn astro add react
yarn astro add react tailwind mdx
# Using PNPM
pnpm astro add react
yarn astro add react tailwind mdx
```

> If your CRA requires any dependencies (e.g. NPM packages), then install them individually using the command line or by adding them to your new Astro project's `package.json` manually and then running an install command. Note that many, but not all, React dependencies will work in Astro. 

Author: I mention what I did above (re: having CRA in title) because of something like this. I don't think we introduced the acronym yet but I don't know how you prefer introducing them / when

### Add your existing app files

Copy your existing Create React App (CRA) project source files and folders (e.g. `components`, `hooks`, `styles`, etc.) into a new folder inside `src/`, keeping its file structure so your app will continue to work. Note that all `.js` file extensions must be renamed to `.jsx` or `.tsx`.

Do not include any configuration files. You will use Astro's own `astro.config.mjs`, `package.json`, and `tsconfig.json`.

Move the contents of your app's `public/` folder (e.g. static assets) into Astro's `public/` folder.

<FileTree>
- public/
  - logo.png
  - favicon.ico
  - ...
- src/
  - cra-project/
    - App.jsx
    - ...
  - pages/
    - index.astro
- astro.config.mjs
- package.json
- tsconfig.json
</FileTree>

### Render your app

Import your app's root component in the frontmatter section of `index.astro`, then render the `<App />` component in your page template:

```astro title="src/pages/index.astro"
---

> import App from '../cra-project/App.jsx'

Reviewer 2: ```suggestion
import App from '../cra-project/App'
```

---
<App client:load/>
```

:::note[client directives]
Your app needs a [client directive](/en/reference/directives-reference/#client-directives) for interactivity. Astro will render your React app as static HTML until you opt-in to client-side JavaScript.

> Use `client:load` to ensure your app loads immediately from the server, or `client:only="react"` to skip server-rendering and run your app entirely client-side.

Reviewer: Nit: although server-side should be hyphened, that doesn't apply to server rendering as far as I know, so I would just change this to follow the SSR name pattern.

```suggestion
Use `client:load` to ensure your app loads immediately from the server, or `client:only="react"` to skip server-side rendering and run your app entirely client-side.
```

:::

> ### Check the Astro config file
> 
> Astro has a configuration file at the root of your project called [`astro.config.mjs`](/en/guides/configuring-astro/). This is used for configuring your Astro project and any installed integrations, including [SSR adapters](/en/guides/deploy/).
> 
> You should see the React integration installed, along with any other official Astro integrations you added.
> 
> ```js
> // astro.config.mjs
> import { defineConfig } from 'astro/config'
> 
> export default defineConfig({
>   integrations: {
>     [react()]
>   }
>   // your configuration options here...
>   // https://docs.astro.build/en/reference/configuration-reference/
> })
> ```
> 

Reviewer 2: ```suggestion
```
Going to remove for now!

See our guide for [configuring Astro](/en/guides/configuring-astro/) for more details and available options.

## Convert your CRA to Astro

> After [adding your existing app to Astro](#add-your-create-react-app-to-astro), you will probably want to convert your app itself to Astro! 

Reviewer: ```suggestion
After [adding your existing app to Astro](#add-your-cra-to-astro), you will probably want to convert your app itself to Astro!
```

> You will replicate a similar component-based design [using Astro HTML templating components for your basic structure](/core-concepts/astro-components/) while importing and including individual React components (which may themselves be entire apps!) for islands of interactivity.

Reviewer: ```suggestion
You will replicate a similar component-based design [using Astro HTML templating components for your basic structure](/en/core-concepts/astro-components/) while importing and including individual React components (which may themselves be entire apps!) for islands of interactivity.
```

> You will replicate a similar component-based design **using Astro HTML templating components for your basic structure** while importing and including individual React components (which may themselves be entire apps!) for islands of interactivity.

Author: I remember you seeing that Astro prefers linking between pages - is the possibility of a link where "using Astro HTML templating components for your basic structure" could be introduced?

I was thinking `/core-concepts/astro-components/` if it hasn't been previously introduced yet in this page.

Reviewer 2: Yeah, I had thought about where exactly to link here (and had included it a tiny bit further down). It is a few paragraphs of plain text in a row, so visually it could totally take another link.  In fact, if we use the link, maybe we don't even need the bold? Maybe that is really what my subconscious was trying to do there!  :smile: Good call, let's try that link and see whether we can get away with no bold!

> Every migration will look different, and can be done incrementally without disrupting your working app. Convert individual pieces at your own pace so that more and more of your app is powered by Astro components over time. 

Reviewer: ```suggestion
Every migration will look different and can be done incrementally without disrupting your working app. Convert individual pieces at your own pace so that more and more of your app is powered by Astro components over time.
```

As you convert your React app, you will decide which React components you will [rewrite as Astro components](#converting-jsx-files-to-astro-files). Your only restriction is that Astro components can import React components, but React components must only import other React components:

```astro title="src/pages/static-components.astro" ins={2,7}
---

> import MyReactComponent from '../components/MyReactComponent.jsx';

Reviewer 2: ``````suggestion
import MyReactComponent from '../components/MyReactComponent';
``````

---
<html>
  <body>
    <h1>Use React components directly in Astro!</h1>
    <MyReactComponent />
  </body>
</html>
```

Instead of importing Astro components into React components, you can nest React components inside a single Astro component:

```astro title="src/pages/nested-components.astro" {2,3,5,8,10}
---

> import MyReactSidebar from '../components/MyReactSidebar.jsx';
> import MyReactButton from '../components/MyReactButton.jsx';

Reviewer 2: ```suggestion
import MyReactSidebar from '../components/MyReactSidebar';
import MyReactButton from '../components/MyReactButton';
```

### Compare: JSX vs Astro

Compare the following CRA component and a corresponding Astro component:

<AstroJSXTabs>
  <Fragment slot="jsx">
    ```jsx title="StarCount.jsx"
    import React, { useState, useEffect } from 'react';
import Header from './Header';
import Footer from './Footer';

const Component = () => {
    const [stars, setStars] = useState(0);
    const [message, setMessage] = useState('');

    useEffect(() => {
        const fetchData = async () => {
            const res = await fetch('https://api.github.com/repos/withastro/astro');
            const json = await res.json();

            setStars(json.stargazers_count || 0);
            setMessage(json.message);
        };

        fetchData();
    }, []);

    return (
        <>
            <Header />
            <p style={{
                backgroundColor: `#f4f4f4`,
                padding: `1em 1.5em`,
                textAlign: `center`,
                marginBottom: `1em`
            }}>Astro has {stars} 🧑‍🚀</p>
            <Footer />
        </>
    )
};

export default Component;
    ```
  </Fragment>
  <Fragment slot="astro">
    ```astro title="StarCount.astro"
    ---

>     import Header from "./header";
>     import Footer from "./footer";

Reviewer 2: ```suggestion
    import Header from "./header.astro";
    import Footer from "./footer.astro";
```

    import "./layout.css";
    const res = await fetch('https://api.github.com/repos/withastro/astro')
    const json = await res.json()
    const message = json.message;
    const stars = json.stargazers_count || 0;
    ---
    <Header />
    <p class="banner">Astro has {stars} 🧑‍🚀</p>
    <Footer />
    <style>
      .banner {
        background-color: #f4f4f4; 
        padding: 1em 1.5em;
        text-align: center;
        margin-bottom: 1em;
      }
    <style>
    ```
  </Fragment>
</AstroJSXTabs>

### Converting JSX files to `.astro` files

Here are some tips for converting a CRA `.js` component into a `.astro` component:

1. Use the returned JSX of the existing CRA component function as the basis for your HTML template.

> 2. Change any [CRA or JSX syntax to Astro](#reference-convert-create-react-app-syntax-to-astro) or to HTML web standards. This includes `{children}` and `className`, for example.

Reviewer: ```suggestion
2. Change any [CRA or JSX syntax to Astro]#reference-convert-cra-syntax-to-astro) or to HTML web standards. This includes `{children}` and `className`, for example.
```

Note that `.astro` and several other file types must be imported with their full file extension.

```astro title="src/pages/authors/Fred.astro"
---
import Card from `../../components/Card.astro`
---
<Card />
```

### CRA Children Props to Astro

Convert any instances of `{children}` to an Astro `<slot />`. Astro does not need to receive `{children}` as a function prop and will automatically render child content in a `<slot />`.

```astro title="src/components/MyComponent" del={3-9} ins={11-13}
---
---
export default function MyComponent(props) { 
    return (
      <div>
        {props.children}
      </div>
    );  
}

<div>
  <slot />
</div>
```

React components that pass multiple sets of children can be migrated to an Astro component using [named slots](/en/core-concepts/astro-components/#named-slots).

See more about [specific `<slot />` usage in Astro](/en/core-concepts/astro-components/#slots).

### CRA Data Fetching to Astro

Fetching data in a Create React App component is similar to Astro, with some slight differences.

> You will need to remove any instances of a side effect hook (`useEffect`) for either `Astro.glob()` or `getCollection()`/`getEntryBySlug()` to access data other files in your project source.

Reviewer: Missing `from`
```suggestion
You will need to remove any instances of a side effect hook (`useEffect`) for either `Astro.glob()` or `getCollection()`/`getEntryBySlug()` to access data from other files in your project source.
```

const data = await response.json();
const randomUser = data.results[0];
---
```

See more about [local files imports with `Astro.glob()`](/en/guides/imports/#astroglob), [querying using the Collections API](/en/guides/content-collections/#querying-collections) or [fetching remote data](/en/guides/data-fetching/).

### CRA Styling to Astro

You may need to replace any [CSS-in-JS libraries](https://github.com/withastro/astro/issues/4432) (e.g. styled-components) with other available CSS options in Astro.

If necessary, convert any inline style objects (`style={{ fontWeight: "bold" }}`) to inline HTML style attributes (`style="font-weight:bold;"`). Or, use an [Astro `<style>` tag](/en/guides/styling/#styling-in-astro) for scoped CSS styles.

```astro title="src/components/Card.astro" del={1} ins={2}
<div style={{backgroundColor: `#f4f4f4`, padding: `1em`}}>{message}</div>
<div style="background-color: #f4f4f4; padding: 1em;">{message}</div>
```

Tailwind is supported after installing the [Tailwind integration](/en/guides/integrations-guide/tailwind/). No changes to your existing Tailwind code are required!

See more about [Styling in Astro](/en/guides/styling/).

## Troubleshooting

> Your CRA might "just work" in Astro! But, it is likely that you may need to make minor adjustments to duplicate your existing app's functionality and/or styles.

Reviewer: Maybe we can reduce this sentence a bit?
```suggestion
Your CRA might "just work" in Astro! But, you may likely need to make minor adjustments to duplicate your existing app's functionality and/or styles.
```
